### PR TITLE
SDK-285 Add experiment to exercise the Feature Variables API

### PR DESCRIPTION
**Pull request by jhugman**
This utilizes Application Services release v77. It adds some quality of life improvements for local development with experiments, 
and an example of using the Variables API. This is only visible: 

 * when no NIMBUS_URL is set
 * accessing the secret developer menu
 * Using the Nimbus Experiments screen to opt in to an experiment. 

Observe the settings title and settings icon in the menu change.

Current state of tooling means that it is possible to enroll in both experiments at the same time. This is only possible with this tooling.

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by jdragojevic** [comment]
Requires a-c to update to appservices 77.0.1.  https://github.com/mozilla-mobile/android-components/pull/10348

File: app/src/main/java/org/mozilla/fenix/experiments/Experiments.kt [hunk 1]
```diff
@@ -4,14 +4,20 @@
...
-    companion object {
-        const val A_A_NIMBUS_VALIDATION = "fenix-nimbus-validation-v3"
-        const val ANDROID_KEYSTORE = "fenix-android-keystore"
-        const val DEFAULT_BROWSER = "fenix-default-browser"
-    }
+/**
+ * Enums to identify features in the app. These will likely grow and shrink depending
+ * on the experiments we want to perform.
+ */
+enum class FeatureId(val jsonName: String) {
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/experiments/Experiments.kt:11]: nit: `jsonName` maybe featureName? Can we add a `@property` to describe that this is :)

File: app/src/main/java/org/mozilla/fenix/experiments/NimbusSetup.kt [hunk 2]
```diff
@@ -66,6 +67,10 @@ fun createNimbus(context: Context, url: String?): NimbusApi =
                 globalUserParticipation = enabled
             }
 
+            if (url.isNullOrBlank()) {
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/experiments/NimbusSetup.kt:70]: Ah nice! Should the entry in `nimbus.remote-settings.url=` still work as well? I am not getting any experiments back locally anymore in a debug build.

File: app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt [hunk 3]
```diff
@@ -5,38 +5,92 @@
...
 
 import mozilla.components.service.nimbus.NimbusApi
-import mozilla.components.support.base.log.logger.Logger
-import org.mozilla.fenix.FeatureFlags
+import org.mozilla.experiments.nimbus.Variables
+import org.mozilla.fenix.experiments.FeatureId
 
-/**
- * Gets the branch of the given `experimentId` and transforms it with given closure.
+/** Gets the branch name of an experiment acting on the feature given `featureId`, and transforms it
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt:11]: nit: This should start on the next line :)

File: app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt [hunk 4]
```diff
@@ -5,38 +5,92 @@
...
+/** The synonym for [getExperimentBranch] to complement [withExperiment(String, (String?) -> T))].
  *
  * Short-hand for `mozilla.components.service.nimbus.NimbusApi.getExperimentBranch`.
  */
-fun NimbusApi.withExperiment(experimentId: String) =
-    this.withExperiment(experimentId, ::identity)
+fun NimbusApi.withExperiment(featureId: FeatureId) =
+    getExperimentBranch(featureId.jsonName)
+
+/** Get the variables needed to configure the feature given by `featureId`.
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt:32]: nit: next line :)

File: app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt [hunk 5]
```diff
@@ -5,38 +5,92 @@
...
+ *
+ * @param featureId The feature id that identifies to the feature under experiment.
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt:34]: nit: .. that identifies the feature..

File: app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt [hunk 6]
```diff
@@ -5,38 +5,92 @@
...
+ *      feature.
+ *
+ * See [recordExposureEvent] for more information on manually recording the event.
+ *
+ * @return a [Variables] object used to configure the feature.
+ */
+fun NimbusApi.getVariables(featureId: FeatureId, recordExposureEvent: Boolean = true) =
+    getVariables(featureId.jsonName, recordExposureEvent)
+
+/** A synonym for `getVariables(featureId, sendExposureEvent)`.
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt:49]: nit: next line.

(same for other methods below)

File: app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt [hunk 7]
```diff
@@ -5,38 +5,92 @@
...
+ *
+ * This exists as a complement to the `withVariable(featureId, sendExposureEvent, transform)` method.
+ *
+ * @param featureId: the id of the feature as it appears in `Experimenter`
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt:53]: nit: remove `:`

(same for other params below)

File: app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt [hunk 8]
```diff
@@ -5,38 +5,92 @@
...
+ * In the case where the use of this function is required, then the `getVariables` function
+ * should be called with `false` so that the exposure event is not recorded when the variables
+ * are fetched.
+ *
+ * This function is safe to call even when there is no active experiment for the feature. The SDK
+ * will ensure that an event is only recorded for active experiments.
+ *
+ *  @param featureId string representing the id of the feature for which to record the exposure
+ *     event.
+ */
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/ext/Nimbus.kt:94]: 👏 Thank you for writing all these docs!

**Review by csadilek — COMMENTED** [review]
Looks great, just some nits and question inline.

File: app/src/main/java/org/mozilla/fenix/experiments/NimbusSetup.kt (hunk 2 above)
Comment (csadilek) [reply app/src/main/java/org/mozilla/fenix/experiments/NimbusSetup.kt:70]: Yes, this still works. @jhugman clarified for me.

**Review by csadilek — APPROVED** [review]
James and I did a round of smoke testing upgrading an existing build to this new version and verified everything works.

**Pull request merged**